network/router.go

```diff
@@ -0,0 +1,394 @@
+			if err := c.Close(); err != nil {
+				log.Error("Couldn't close secure connection:",
+					err)
+			}
+			return
+		}
+		// pass it along
+		r.launchHandleRoutine(dst, c)
+	})
+	if err != nil {
+		log.Error("Error listening:", err)
+	}
+}
+
+// Stop will stop the listening routine, and stop any routine of handling
+// connections. Calling r.Start(), then r.Stop() then r.Start() again leads to
+// an undefined behaviour. Callers should most of the time re-create a fresh
+// Router.
+func (r *Router) Stop() error {
+	var err error
+	if r.host.Listening() {
+		err = r.host.Stop()
+	}
+	err2 := r.stopHandling()
+	r.reset()
+	if err != nil {
+		return err
+	} else if err2 != nil {
+		return err2
+	}
+	return nil
+}
+
+// Send sends to an ServerIdentity without wrapping the msg into a SDAMessage
+func (r *Router) Send(e *ServerIdentity, msg Body) error {
+	if msg == nil {
+		return errors.New("Can't send nil-packet")
+	}
+
+	c := r.connection(e.ID)
+	if c == nil {
+		var err error
+		c, err = r.connect(e)
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Lvlf4("%s sends to %s msg: %+v", r.address, e, msg)
+	var err error
+	err = c.Send(msg)
+	if err != nil {
+		log.Lvl2(r.address, "Couldn't send to", e, ":", err, "trying again")
+		c, err := r.connect(e)
+		if err != nil {
+			return err
+		}
+		err = c.Send(msg)
+		if err != nil {
+			return err
+		}
+	}
+	log.Lvl5("Message sent")
+	return nil
+}
+
+func (r *Router) connect(si *ServerIdentity) (Conn, error) {
+	c, err := r.host.Connect(si.Address)
+	if err != nil {
+		return nil, err
+	}
+	if err := r.negotiateOpen(si, c); err != nil {
+		return nil, err
+	}
+
+	r.launchHandleRoutine(si, c)
+	return c, nil
+
+}
+
+// handleConn is the main routine for a connection to wait for incoming
+// messages.
+func (r *Router) handleConn(remote *ServerIdentity, c Conn) {
+	defer func() {
+		// when leaving, unregister the connection
+		if err := c.Close(); err != nil {
+			log.Error(r.address, "having error closing conn to ", remote.Address, ":", err)
+		}
+		// and release one on the waitgroup
+		r.wg.Done()
+	}()
+	address := c.Remote()
+	log.Lvl3(r.address, "Handling new connection to ", remote.Address)
+	for {
+		packet, err := c.Receive()
+		// So the receiver can know about the error
+		packet.SetError(err)
+		packet.From = address
+		packet.ServerIdentity = remote
+
+		// whether the router is closed
+		if r.Closed() {
+			// signal we are done with this go routine.
+			return
+		}
+
+		if err != nil {
+			// something went wrong on this connection
+			log.Lvlf4("%+v got error (%+s) while receiving message", r.id.String(), err)
+
+			if err == ErrClosed || err == ErrEOF || err == ErrTemp {
+				// remote connection closed
+				log.Lvl3(r.address, "handleConn with closed connection: stop (dst=", remote.Address, ")")
+				return
+			}
+			// weird error let's try again
+			log.Lvl3(r.id, "Error with connection", address, "=>", err)
+			continue
+		}
+
+		if err := r.Dispatch(&packet); err != nil {
```

Comment:
It's maybe only me it seems, but instead of having hierarchies of `if` inside `if` etc where the "normal" behavior happens *inside* the if, I prefer that the `if` checks if something is wrong, and the normal behavior stays at the most possible top level indentation so it's way more easy to follow. 
Same story here, I could put the `r.Dispatch` inside an `else` but then,  the `Dispatch` which is the **normal behavior** is not emphasized, rather put alongside a big `if` which checks if something is wrong.
Most of the times it's like that in the wild, for example, the first example to me seems cleaner than the second:
```
f, err := os.Open("filename.ext")
if err != nil {
    log.Fatal(err)
}
// do something with the open *File f
// then something else
```
vs:
```
f, err := os.Open("filename.ext")
if err != nil {
    log.Fatal(err)
} else {
    // do something with the open *File f
}
// then something else
```
Yeah sorry again, I put up a large comment ... :D